R/matching.R

```diff
@@ -157,3 +157,99 @@ common <- function(x, table, tolerance = Inf,
                    duplicates = c("keep", "closest", "remove")) {
     !is.na(closest(x, table, tolerance = tolerance, duplicates = duplicates))
 }
+
+#' @rdname matching
+#'
+#' @details
+#' `join`: joins two `numeric` vectors by mapping values in `x` with
+#' values in `y` and *vice versa* if they are similar enough (provided the
+#' `tolerance` and `ppm` specified). The function returns a `matrix` with the
+#' indices of mapped values in `x` and `y`. Parameter `type` allows to define
+#' how the vectors will be joined: `type = "left"`: values in `x` will be
+#' mapped to values in `y`, elements in `y` not matching any value in `x` will
+#' be discarded. `type = "right"`: same as `type = "left"` but for `y`.
+#' `type = "outer"`: return matches for all values in `x` and in `y`.
+#' `type = "inner"`: report only indices of values that could be mapped.
+#'
+#' @param y `numeric`, the values to be joined. Should be sorted.
+#' @param ppm `numeric(1)` representing a relative, value-specific
+#'  parts-per-million (PPM) tolerance that is added to `tolerance`.
+#' @param type `character(1)`, defines how `x` and `y` should be joined. See
+#' details for `join`.
+#'
+#' @note `join` is based on `closest(x, y, tolerance, duplicates = "closest")`.
+#' That means for multiple matches just the closest one is reported.
+#'
+#' @return `join` returns a `matrix` with two columns, namely `x` and `y`,
+#' representing the index of the values in `x` matching the corresponding value
+#' in `y` (or `NA` if the value does not match).
+#'
+#' @export
+#' @examples
+#'
+#' ## Join two vectors
+#' x <- c(1, 2, 3, 6)
+#' y <- c(3, 4, 5, 6, 7)
+#'
+#' jo <- join(x, y, type = "outer")
+#' jo
+#' x[jo[, "x"]]
+#' y[jo[, "y"]]
+#'
+#' jl <- join(x, y, type = "left")
+#' jl
+#' x[jl[, "x"]]
+#' y[jl[, "y"]]
+#'
+#' jr <- join(x, y, type = "right")
+#' jr
+#' x[jr[, "x"]]
+#' y[jr[, "y"]]
+#'
+#' ji <- join(x, y, type = "inner")
+#' ji
+#' x[ji[, "x"]]
+#' y[ji[, "y"]]
+join <- function(x, y, tolerance = 0, ppm = 0,
+                 type = c("outer", "left", "right", "inner")) {
+    switch(match.arg(type),
+           "outer" = .joinOuter(x, y, tolerance = tolerance + ppm(x, ppm)),
+           "left" = .joinLeft(x, y, tolerance = tolerance + ppm(y, ppm)),
+           "right" = .joinRight(x, y, tolerance = tolerance + ppm(x, ppm)),
+           "inner" = .joinInner(x, y, tolerance = tolerance + ppm(x, ppm))
+    )
+}
+
+.joinLeft <- function(x, y, tolerance) {
+    cbind(x = seq_along(x),
```

Comment:
Actually, I don't see the benefit from returning a `matrix` instead of a `list` of two `integer`. The `list` will always be faster both when creating:
```r
> microbenchmark(cbind(1:4, 1:4), list(1:4, 1:4))
Unit: nanoseconds
            expr  min     lq    mean median     uq   max neval cld
 cbind(1:4, 1:4) 1214 1256.5 1550.00   1324 1492.0 16353   100   b
  list(1:4, 1:4)  342  359.5  531.43    374  474.5 10706   100  a 
```

as well when extracting/using:
```r
> mat <- cbind(1:10, 1:10)
> lst <- list(y = 1:10, y = 1:10)
> microbenchmark(mat[, 1], lst$x)
Unit: nanoseconds
     expr min  lq   mean median    uq   max neval cld
 mat[, 1] 516 528 745.95  537.5 561.5 18541   100   b
    lst$x 262 272 356.91  286.5 294.5  5940   100  a 
```

I usually also don't like returning `list` from a function, but since this is supposed to be a function that will be used by us in other packages but less frequently (if at all) by *normal* users I would opt for the faster implementation (i.e. `list`).

Reply:
Thanks for this example. While I don't like returning lists either you convinced me to do it in this special use case.

> since this is supposed to be a function that will be used by us in other packages but less frequently (if at all) by *normal* users I would opt for the faster implementation

If that is the case we could keep the name `join` because we won't use `tibble`s or the `tidyverse` internal anyway. :wink:

Reply:
😄